### PR TITLE
Support circular layout barplots; show legends for barplots using length-scaling

### LINKED ISSUE #297: Add support for circular layout barplots
Continuation of #201, since the #293 PR only supports rectangular layouts.

### LINKED ISSUE #343: Don't immediately draw barplots when clicking the "Draw barplots?" checkbox, to avoid crashing
This behavior isn't a problem when there's feature metadata provided (and the default barplot layer is just an empty red bar or something), but when there isn't feature metadata -- and barplots default to stacked barplots of sample metadata -- then the current behavior means that a stacked barplot of whatever the "first" sample metadata field is will be drawn. If the sample metadata field in question is not categorical and/or it has a ton of groups, then the resulting stacked barplot will take a super long time to draw and/or just straight-up crash the browser.

This is currently a problem with the EMP tree (~200k tips with the chunk of it I'm testing with): the first sample metadata field is `adiv_chao1`, and trying to draw a stacked barplot of that just straight up crashes the page in Chromium. (I get a [`SIGTRAP` error](https://en.wikipedia.org/wiki/Signal_(IPC)#SIGTRAP), bizarrely enough. Didn't know what that meant.).

By deferring drawing barplots until the user presses "Update", the user is responsible for choosing what field to use. (I guess this sort of matches the behavior Emperor plots with hundreds of thousands of samples -- IIRC trying to color by a similarly silly field there will also hang, but when the hang happens it's clear to the user _why_ things hanged.)

Changing this in the code should really just boil down to removing [this line](https://github.com/biocore/empress/blob/4f6ddd3a4e03b85703f733bd75b9771ef82157b4/empress/support_files/js/barplot-panel-handler.js#L77) of code, but we'll have to update the README as well since it mentions how a red bar pops up immediately (rather than on pressing Update).

### LINKED ISSUE #354: Show a legend for barplot layers using length scaling
This would just show the min / max value of the field used for length scaling. It'd be a nice way to convey some extra information to the user, and could help with determining what an adequate min / max length for the visualization should be: for example, in the MP dataset `Confidence` field, the min value is ~0.688 (with the max being ~0.999), so it makes sense to set the minimum length to 69 if the max length is 100 (rather than using the default min length of 1).

I already have a demo of this ready on a branch continuing from #340 (see screenshot below). However, it required some refactoring, so I think it'd make sense to merge #340 in first and then add this in when it's ready in a few days.

fancyleg

### PR DESCRIPTION
**Pull request by fedarko**
This PR replaces #346 (this is the branch I ended up resolving the conflicts in). It closes #354 (simple "legends" are now shown for all barplot layers using length scaling, showing the min/max values in a field), and since this includes all of #346's stuff this also closes #297 (circular layout barplots) and #343 (make the user press `Update` to draw barplots, rather than having the `Draw barplots` checkbox immediately show barplots).

fancy

Copying the text of #346 and its comments below. Many of the points brought up are still relevant (barplot lengths are still arbitrary, etc.).

---

Closes #297 and closes #343.

This should probably not be merged until #345 and #340 are, but this branch does work in case people want to use this functionality.

Thanks!

### Issues we should probably discuss

- Barplot lengths are still in arbitrary units, so a barplot with length "100" looks thinner in the moving pictures circular layout than in the rectangular layout. Standardizing these somewhat should be possible, but I wanted to avoid doing that until #345 is in (to avoid having to redo the work to accommodate JS layouts).

- Circular layout bars are drawn here by drawing two rectangles (reusing an image I posted [here](https://github.com/biocore/empress/issues/297#issuecomment-676949805)) --
  image
  We may want to smooth this out, and look at the number of tips in the tree to determine how many rectangles to draw (e.g. use ~15 rectangles for tiny trees like the one above, and only 1 rectangle for massive EMP-scale trees). Or we could just leave this as is until the future, when we'd ideally have a shader draw these as actual curved sections.

- Although the functions that add the coordinates for individual bars (in both rect/circular layouts) are now tested, `addFMBarplotLayerCoords()` and `addSMBarplotLayerCoords()` are still untested. Testing these is going to be a bit time-consuming -- testing the trig stuff gets really annoying really quickly (see e.g. the `_addCircularBarCoords()` test here...), so I think it makes sense to defer testing these until we agree on how many rectangles to use for approximating a curve (per the above point).

Comments from #346
---

> Barplot lengths are still in arbitrary units, so a barplot with length "100" looks thinner in the moving pictures circular layout than in the rectangular layout. Standardizing these somewhat should be possible, but I wanted to avoid doing that until #345 is in (to avoid having to redo the work to accommodate JS layouts).

Is it possible to use a fraction of the spatial length of an "average branch" as a unit? This could then be computed per layout and things should be roughly equivalent at least relative to the tree.

You could for example find the mean branch length, and then find whatever branch is closest to that value, look at the euclidean distance of the vertices (from the layout data) for that branch and take a fraction of that value. Maybe I'm missing something but it seems like what we need is a frame of reference.

_Originally posted by @ElDeveloper in https://github.com/biocore/empress/pull/346#issuecomment-678555741_

---

> > Barplot lengths are still in arbitrary units, so a barplot with length "100" looks thinner in the moving pictures circular layout than in the rectangular layout. Standardizing these somewhat should be possible, but I wanted to avoid doing that until #345 is in (to avoid having to redo the work to accommodate JS layouts).
> 
> Is it possible to use a fraction of the spatial length of an "average branch" as a unit? This could then be computed per layout and things should be roughly equivalent at least relative to the tree.
> 
> You could for example find the mean branch length, and then find whatever branch is closest to that value, look at the euclidean distance of the vertices (from the layout data) for that branch and take a fraction of that value. Maybe I'm missing something but it seems like what we need is a frame of reference.

I think this could work. This boils down to the same problem as with line width (#276) -- I think trying to see how "long" a given node  of a known length is in one layout, and then using that length somehow when scaling barplot lengths / line widths, could work.

Actually -- thinking about this some more ... So our goal here is making sure that the drawn length (in terms of "outer radius" - "inner radius" of a barplot layer with length L is proportional to the drawn length (in terms of "rightmost x - innermost x") of another barplot layer in the rectangular layout with the same length. These measures both only involve a single "axis" (x in the rectangular layout, radius in the polar coordinates for the circular layout). What we could do is use the drawn length of the [maximum-displacement node](https://github.com/fedarko/empress/blob/7b88512e351ebb15c656ed648719fcd3c570b80a/empress/support_files/js/empress.js#L1248-L1287) (the node with the rightmost/max x coordinate for the rectangular layout, the node with the max radius for the circular layout) as a "unit" for drawing bars -- so maybe a user-specified length of "100" is equal to 1/10 of that length, or something similar. ...Maybe?

I'll think about this more next week. I have a suspicion that this scaling stuff isn't really that bad, but every time I try to think about it my brain goes to mush :P

_Originally posted by @fedarko in https://github.com/biocore/empress/pull/346#issuecomment-678580565_

### CODE REVIEW

**Comment by emperor-helper** [comment]
The following artifacts were built for this PR: [empire-biplot.qzv](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fmchelper.ucsd.edu%2Fdownloads%2Fempress%2F357%2Fempire-biplot.qzv), [empire.qzv](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fmchelper.ucsd.edu%2Fdownloads%2Fempress%2F357%2Fempire.qzv), [empress-tree.qzv](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fmchelper.ucsd.edu%2Fdownloads%2Fempress%2F357%2Fempress-tree.qzv)

File: tests/test-empress.js [hunk 1]
```diff
@@ -1026,5 +1026,310 @@ require(["jquery", "UtilitiesForTesting", "util", "chroma"], function (
...
+            // Note that bL and bR remain consistent throughout all of the
+            // triangles drawn: these are always at the same angle as the node
+            // this bar is being drawn for.
+            //
+            // We consider "left" positions as those using the outer radius,
+            // and "right" positions as those using the inner radius.
+            // (Of course when you look at an opposite side of the circle left
+            // and right'll be switched around. We just stick to this notation
+            // here to make describing and testing this less difficult.)
+            _.each(coords, function (v, i) {
```
Comment (ElDeveloper) [inline tests/test-empress.js:1205]: Would you mind providing a bit more intuition for why go with this approach for testing rather than writing out the expected values for the elements in the arrays?

**Comment by fedarko** [comment]
> I just noticed something a bit off with the bar scaling, maybe worth fixing here? When you click on the scale checkbox and you have a length different to 100 (the default), the maximum length is set by default to 100. I think this should by default be the maximum length so that bars don't shrink out of the blue.

That's a really good point, and I agree that the UI for this is a bit confusing. Part of the challenge here is that, if the user has already done some length-scaling stuff for a layer, I'm not sure we want to mess with that when the user switches back to non-length-scaling -- do you think an acceptable compromise would be setting the "max length" for scaling to the default length the first time the "length scaling" checkbox is checked? Or maybe we should just update it every time.

I guess in the same vein we should definitely synchronize the default lengths between feature and sample metadata barplots --

asdf

I'll open up a new issue for this if that sounds ok with you; I think fixing this would be mildly out of scope for this PR.

File: tests/utilities-for-testing.js [hunk 2]
```diff
@@ -285,24 +285,58 @@ define(["Empress", "BPTree", "BiomTable"], function (
...
-        deepEqual(toFixedIfy(arr1), toFixedIfy(arr2), message);
+    /**
+     * Calls stringify() on two numbers and calls deepEqual() on the result.
+     *
+     * @param {Number} n1
+     * @param {Number} n2
+     * @param {String} message (optional)
+     */
+    function approxDeepEqual(n1, n2, message) {
+        deepEqual(stringify(n1), stringify(n2), message);
```
[the remark this answers is not in the document]
Comment (fedarko) [reply tests/utilities-for-testing.js:321]: I guess subtracting and checking that the difference is less than some epsilon would be a clearer way to do this. The original reason I added this stuff in IIRC was as a quick hack to fix the circular layout JS tests that spiraled out of control; e.g. `Math.cos(Math.PI / 2))` is something like `6.1232e-17` and not `0`, which was breaking `===`. Converting to strings (rather than doing more floating-point math) seemed like the easier option.

Thinking about it some more, since we're using `.toFixed(4)` to convert numbers to strings, this should really just behave identically\* to subtracting numbers and checking that the difference is smaller than `0.0001`. So I don't _think_ replacing the stringify stuff with subtraction would add much besides arguably being a bit clearer.

Anyway, long story short, I can rip the stringify stuff out and replace it with something more conventional if you'd like me to. Just let me know :)

\* I hesitate to say "identically" here considering how bizarre JS' type system is, but if we assume the inputs are Numbers it should be close enough

File: empress/support_files/js/empress.js [hunk 3]
```diff
@@ -857,6 +857,175 @@ define([
...
+                angleCos: angleCos,
+                angleSin: angleSin,
+                lowerAngleCos: lowerAngleCos,
+                lowerAngleSin: lowerAngleSin,
+                upperAngleCos: upperAngleCos,
+                upperAngleSin: upperAngleSin,
+            };
+        } else {
+            // When layouts are ported to JS in the future, catching this error
+            // will be more important -- since then the node's angle
```
[the remark this answers is not in the document]
Comment (fedarko) [reply empress/support_files/js/empress.js]: oops, that is indeed an artifact of the distant past of a few days ago :sweat_smile: should be updated now.

File: empress/support_files/js/empress.js [hunk 4]
```diff
@@ -1137,55 +1307,83 @@ define([
...
+                    // radius value in the layout? Not sure if this step is a
+                    // bottleneck worth spending time working on, though.)
+                    var r =
+                        this.getX(node) /
+                        Math.cos(this.getNodeInfo(node, "angle"));
+                    if (r > maxD) {
+                        maxD = r;
+                    }
+                } else {
+                    throw new Error("Unsupported barplot layout");
```
[the remark this answers is not in the document]
Comment (fedarko) [reply empress/support_files/js/empress.js]: Agreed, checking the layout every time through the loop is silly. Just updated things so the function to use for comparing `maxD` and a node's x / radius is set beforehand.

File: tests/utilities-for-testing.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ElDeveloper) [reply tests/utilities-for-testing.js:321]: My suggestion would be to not add more surprises to the already surprising world that JavaScript is. Can you replace these string comparisons for arithmetic-based comparisons?

**Comment by ElDeveloper** [comment]
> I guess in the same vein we should definitely synchronize the default lengths between feature and sample metadata barplots --

Thinking more about this. I think this is only a big problem for this tree that has an unusually long branch. Let's table this until after we sort out consistent lengths, etc

File: tests/test-empress.js (hunk 1 above)
Comment (fedarko) [reply tests/test-empress.js:1205]: Honestly, it's exactly what it looks like. I _really_ didn't want to write out all 61 elements manually, especially since the calculations were going to involve trigonometry functions (resulting in either numbers that are incomprehensible to someone reading the code, or me writing out something like `200 * Math.cos(-piover2)` repeatedly); hence this kind-of-too-much-work-for-its-own-good solution.

I mean, it works now, and I think generalizing this solution to handle changes in the `coords` formatting (e.g. r/g/b being compressed into a single number) would actually be pretty simple. So if you don't mind I would prefer to keep this as is for the time being, rather than rewrite it to check things explicitly (especially since the explicit method would probably need un-breaking in the near future).

File: tests/utilities-for-testing.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (fedarko) [reply tests/utilities-for-testing.js:321]: Sounds good, will do.

File: tests/test-empress.js (hunk 1 above)
Comment (ElDeveloper) [reply tests/test-empress.js:1205]: That makes sense - thanks for the explanation!

File: tests/utilities-for-testing.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ElDeveloper) [reply tests/utilities-for-testing.js:321]: Thanks!

File: tests/test-empress.js (hunk 1 above)
Comment (kwcantrell) [reply tests/test-empress.js:1205]: It would be nice if we could avoid this approach as it make following the test a bit harder since we have to following a bunch of switch statements/function calls. Perhaps a better approach would be to create a small tree, like 2 or 3 tips, and then manually set the location of the tips in places that make the trigonometric functions easier to calculate such as (1,0) and (-1,0) (i.e. places s.t. sin/cos would be equal to +-1 or 0).  However, I will leave that up to you and whether or not you would want to change the test in this PR.

Also. keep in mind that x,y and color are no longer going to be in the same array buffer. So regardless of the approach we use, we are going to have to modify it.

File: empress/support_files/js/empress.js [hunk 5]
```diff
@@ -857,6 +857,175 @@ define([
...
     };
 
+    /**
+     * Returns an Object describing circular layout angle information for a
+     * node.
+     *
+     * @param {Array} node An array that holds all the attributes for a given
+     *                    node. As with the "node" parameter of
+     *                    this.getNodeInfo(), this is an entry in the tree
+     *                    data.
```
[the remark this answers is not in the document]
Comment (fedarko) [reply empress/support_files/js/empress.js]: Oops, good catch! Yeah, I had updated the places where this function was called to just have `node` be the postorder position, but I forgot to update the docs here. Fixed now.

File: empress/support_files/js/empress.js [hunk 6]
```diff
@@ -1222,19 +1427,24 @@ define([
...
+     *                              changed to provide length-scaling legend
+     *                              information, as is done in
+     *                              addFMBarplotLayerCoords(); however for now
+     *                              that isn't supported.)
      */
     Empress.prototype.addSMBarplotLayerCoords = function (
         layer,
         coords,
-        prevLayerMaxX
+        prevLayerMaxD
```
[the remark this answers is not in the document]
Comment (fedarko) [reply empress/support_files/js/empress.js:1473]: Whoops, yeah, I forgot to update the docs for this. Should be good now. Thanks for catching this!

File: empress/support_files/js/empress.js [hunk 7]
```diff
@@ -1137,55 +1307,83 @@ define([
...
-        var maxX = -Infinity;
+        // The "D" stands for displacement. For the rectangular layout, this
+        // variable represents the rightmost node's x-coordinate; for the
+        // circular layout, this variable represents the node farthest away
+        // from the root's distance from the root (a.k.a. radius, since the
+        // root is (0, 0)).
+        // In any case, the maxD value is used when determining how "close"
+        // to the root node we can start drawing barplots without crossing over
+        // any nodes.
+        var maxD = -Infinity;
```
[the remark this answers is not in the document]
Comment (fedarko) [reply empress/support_files/js/empress.js]: Makes sense. This is now computed in its own function -- only thing is right now i'm calling it in `getLayoutInfo()` rather than `updateLayout()` just to make sure that the coordinate info is always consistent with the max displacement.

File: tests/utilities-for-testing.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (fedarko) [reply tests/utilities-for-testing.js]: OK, `approxDeepEqual()` and `approxDeepEqualMulti()` now work by checking the absolute difference of two numbers! The epsilon is configurable; it defaults to `1e-5` for now.

**Comment by fedarko** [comment]
OK, all comments should be addressed now. Thanks @kwcantrell and @ElDeveloper!

**Comment by fedarko** [comment]
Just a quick update -- I noticed that there's a bug in the master version of Empress where searching for an invalid node name wouldn't change the styling of the search bar and would instead throw an error:

image

It turns out this is because `BPTree.getNodesWithName()` returns an empty array (`[]`) if no nodes match a name, rather than `undefined` (which is what `CanvasEvents.placeNodeSelectionMenu()` expected). I've pushed a fix to this to this PR, in af0bd44 and 28e2bae.

File: empress/support_files/js/empress.js [hunk 8]
```diff
@@ -203,6 +203,19 @@ define([
          */
         this._yrscf = null;
 
+        /**
+         * @type {Number}
+         * For the rectangular layout (assuming no rotation has been done),
```
Comment (kwcantrell) [inline empress/support_files/js/empress.js:208]: It might be better to remove the note about rotations. "Rotations" will be handled by rotating the virtual camera which means empress will have no idea the tree is actually being rotated. But I'll leave that decision up to you.

**Comment by fedarko** [comment]
Thanks @ElDeveloper and @kwcantrell! Comments have been addressed, and I merged in the #360 changes now in master.

**Comment by kwcantrell** [comment]
Thanks @fedarko!

**Comment by ElDeveloper** [comment]
Thanks both!
